### ISSUE TITLE
Large Language Models Cannot Self-Correct Reasoning Yet

### ISSUE THREAD

**Issue by dreamingjudith** [dreamingjudith/paper-reading-list#36]
[Arxiv](https://arxiv.org/abs/2310.01798)

## Introduction

- LLM에 대해 accuracy, reasoning capabilites, safety of their generated content에 대한 염려
- 이에 대한 해결책으로 급부상 중인 LLM의 Self-correction https://arxiv.org/pdf/2210.11610.pdf
  - "LLMs refine their responses based on feedback to their previous outputs"
  - However, the underlying mechanics and efficacy of self-correction in LLMs remain underexplored.
  - 의문점: If an LLM possesses the ability to self-correct, why doesn’t it simply offer the correct answer in its initial attempt?
- Intrinsic self-correction
  - "a scenario wherein the model endeavors to rectify its initial responses based solely on its inherent capabilities, without the crutch of external feedback"
  - 현실 세계에서도 고품질의 외부 피드백을 받을 수 없는 상황이 많이 있기 때문에 상당히 중요한 가정이긴 함
  - 근데 막상 연구해보니 기존 연구들과는 달리 LLM이 external feedback이 없는 환경에서는 이전 응답을 고치기 힘들어 하고, 많은 경우 오히려 고치고 난 후가 이전보다 성능이 떨어졌음
  - "we observe that the improvements in these studies result from using oracles to guide the self-correction process, and the improvements vanish when oracle labels are not available"
  - 대신 Multi-agent debate*의 잠재력을 조사해봄  
  \* multiple instances of an LLM critique each other’s responses
  - 근데 Multi-agent debate도 응답 수를 동일하게 맞춘다는 조건 하에서는 Self-consistency*보다 더 나을 게 없었다  
  \* https://arxiv.org/abs/2203.11171
- 근데 또 Self-correction이 응답의 스타일 변화나 적절성을 개선하는 작업에는 효과적이었음
- 그래서 이를 좀 더 깊이 이해하기 위해 Self-correction을 일종의 post-hoc prompting의 관점에서 보기로 함 (이게 뭔 소리냐면 이 때의 prompting은 LLM의 응답을 포함하고 있기 때문)
- Self-correction이 정말로 도움이 되는 시나리오를 평가함 (특히 보편적인 프롬프트가 없는 상황에서 LLM이 instruction이나 feedback을 줄 때)
- 더구나 특정 태스크에서 Self-correction이 가져온 향상은 처음부터 잘못 만들어진 instruction이지만 상당히 정교하게 만들어진 피드백 프롬프트에 의해 문제점이 가려져 있는 경우도 있었음을 발견함. (이런 경우는 피드백을 최초 instruction부터 적용하거나 아니면 initial prompt를 더 좋게 만듦으로써 성능을 향상하거나 inference cost를 낮출 수 있음.)
- 결국 이 논문 통해서 하고 싶은 말은 Self-correction에 대해 좀 비판적으로 접근할 필요가 있으며 external feedback과 같은 방법을 고안하기를 바란다

## Background and Related Work

- 생각해보면 인공 신경망의 학습의 기본은 prediction error에 따라 model parameter를 조정하는 거였는데 이게 결국 training data로부터 feedback	을 받아 오류를 수정하는 것으로 볼 수 있음
- 강화학습도 Self-correction의 다른 지평. 그래서 최근 LLM에서 주목할 만한 발전을 이룩한 것이 RLHF(Reinforcement Learning from Human Feecback)을 적용한 사례였음
- 일반적인 강화학습은 온전히 environmental feedback에만 의존하지만, RLHF는 인간이 가지고 있는 통찰력을 활용하기 때문에 좀 더 정확한 reward function의 설계가 가능하며 의도하지 않은 행동을 줄일 수 있음
- 그러나 RLHF는 training phase의 피드백이기 때문에 실시간성은 떨어지며, 이 때문에 inference 중 특정 쿼리에 대해서는 correction이 동작하지 않을 수도 있음
- LLM의 발전에 따라 inference-time self-correction의 가능성이 주목을 받으면서 이제는 self-correction에 대해 모델이 자신의 출력에 대해 얼마나 적절한지 혹은 정확한지를 인지할 수 있는가, 혹은 필요한 경우 정답을 수정할 수 있는가를 보기 시작했다
- 근데 사실 Self-correction이라는 개념 자체도 문헌마다 다르게 정의되고 있으며, 그러다보니 유사 개념으로 쓰이고 있는 "Self-refine", "Self-critique", "Self-improve" 등등도 다 저마다의 문맥을 가지고 있음
- 이런 개념들을 사용할 때 가장 중요한 건 피드백의 소스가 어디냐임
  - Purely internal: originating solely from the LLM
  - External inputs: from humans, other models, or external tools and knowledge sources
- 보통은 고품질의 외부 피드백이 불가능한 상황이 많으며, 또 외부 피드백 자체가 넘사벽의 이점이기 때문에 본 연구에서는 LLM이 스스로 응답을 고칠 능력을 내재하고 있는지 평가하고 이를 통해 LLM의 능력치를 이해하고자 함

#### Experimental setup

- Benchmarks
  - GSM8K
  - CommonSenseQA
  - HotpotQA
- Prompts: 3-step prompting strategy
  1. Initial generation (as standard prompting)
  2. Reviewing its previous generation and producing feedback
  3. Propmting the model to answer the original question again with the feedback
- Test models
  - GPT-3.5-Turbo
    - 선행 논문의 환경 그대로를 재현하려고 했지만 상세 스펙이 없어 gpt-3.5-turbo-0613을 사용
    - 0301의 경우 특히나 선행 논문의 프롬프트를 이용할 때 최초 응답부터 거의 정확했었음
  - GPT-4 accesed on 2023/08/29
    - 최신 모델이며 성능이 가장 좋다고 해서 자체적으로 실험
    - 비용 절감을 위해 랜덤으로 200개씩만 추출해서 질의함 (HotpotQA는 100개)
- Other setup
  - 기본 temperature값으로 설정하고 최대 2라운드를 거쳐 Self-correction을 수행
  - 선행 논문대로 정답 레이블을 가지고 언제 self-correction loop을 멈춰야 하는지를 판단함

#### Results

선행 논문에서 보고했던 결과와 유사한 수치가 나옴

#### Reflection

- 근데 생각해보면 현실적으로 언제 Self-correction을 멈춰야 하는지를 알 수가 없음 (수학 문제 풀라고 시켰을 때 암산왕이 아닌 이상 모델 출력만 보고 더 고쳐도 되는지 아닌지 판단할 수 없는 것처럼)
- 그래서 베이스라인으로 random guessing을 설정함
  - 언제 멈춰야 하는지를 판단하기 위해 정답 레이블을 이용하는 건 그대로임
  - 근데 고치는 걸 LLM한테 시키는 게 아니라 남은 옵션 중에서 랜덤으로 추정하게 함
  - 뭔 소리냐면 CommonSenseQA의 경우 5지선다 문제이기 때문에 정답 하나 고르고 나도 4개의 선택지에 대해서도 생성을 해볼 수 있게 된다는 것
- 그랬더니 CommonSenseQA에서 한 4라운드 돌려보면 Random guessing의 정확도는 100%를 찍어버리게 된다. Self-correct에 비해 월등히 좋음
- 생각해보면 애초에 정답지를 갖고 있는 상태에서 정답과 유사해질 때까지 모델을 인퍼런스 돌려보는 실험 자체가 무슨 의미가 있나 싶긴 함
  - 코드 생성과 같은 태스크에서는 만들어진 코드로 unit test를 돌렸을 때 통과하냐 안 하냐가 정확도의 기준이 되기 때문에 정답을 미리 알고 있다고 가정할 수 있음
  - 그러나 수학 문제 풀기와 같은 reasoning task에서는 정답을 이미 알고 있다는 것 자체가 "태스크를 수행한다" 라는 의미를 상실하게 만들어버린다는 것
- 하지만 이러한 시도 자체가 LLM의 reasoning process를 분석하기 위함이라고 생각하면, 좀 더 구체적인 reasoning 과정을 분석하거나 모델이 self-correction할 때마다 내는 다양한 솔루션을 탐색해볼 수 있지 않냐는 것

#### Why Does This Performance Not Increase, but Instead Decrease?

- 그래서 앞서 Introduction에서 정의했던 것처럼 correct label 없이 무조건적으로 2라운드씩 Self-correction을 수행해봤다
- 그 결과 모든 벤치마크에서 성능이 떨어짐
- 왜?
  - Empirical analysis
    - 통계 분석을 해봤더니 GSM8K의 경우 응답의 74.7%는 변화가 없는 가운데 오답을 정답으로 다시 고친 경우가 7.6%에 불과함
    - CommonSenseQA의 경우 더 처참하다. 무려 39.8%가 정답을 오답으로 고침. 왜 그런가 하고 봤더니 문제 자체가 5지선다형이라 Self-correction을 시키면 모델이 다른 걸 고르게끔 유도하는 효과가 있었다
    - 아까 Oracle 기반의 실험에서는 Self-correction 결과가 맞는 답을 오히려 오답으로 고치지 않게끔 Correct label을 활용했는데, 사실 "어떻게 정답을 오답으로 고치지 않게 막을 것이냐"는 문제가 Self-correction의 핵심이라고 볼 수 있음
  - Intuitive explanation
    - Initial prompt만으로 정답이 나왔다면 그 시점의 프롬프트 디자인이나 모델의 디코딩 알고리즘이 맞아 떨어졌다는 건데, 그 상황에서 additional input을 주게 되면 오히려 모델을 잘못된 방향으로 틀어버리는 꼴이 됨

#### Results of Other Prompts and Related Studies

- Self-correction prompt가 suboptimal했다고 하면, 다른 프롬프트를 썼을 때 성능이 더 오른다는 소리네?
  - 물론 특정 벤치마크별로 성능 개선을 이끄는 프롬프트가 뭔지는 사람이 판단할 수는 있다
  - 근데 그런 식의 접근은 이전에 true few-shot setting 논문에서 논의했던 것과 비슷한 흐름으로 더이상 우리가 정의한 intrinsic self-correction이 아니게 되어버림
  - 그래서 다른 프롬프트들을 이용해서 돌려봤는데 역시 성능은 오르지 않았다
- "성능을 높여주는 self-correction prompt가 있는가?"라는 질문 대신 "Are large language models really able to self-correct their reasoning based solely on their inherent capabilites?"에 초점을 맞추자

### Multi-Agent Debate and Self-Consistency

- Multi-agent debate: 모델 호출을 여러 번 해서 마치 여러 모델이 서로 토론하는 것처럼 꾸미는 기법으로, 실제 Reasoning task에서 유의미한 성능 향상이 있었다고 함
- 실험 세팅
  - 선행 논문과 동일한 프롬프트를 가지고 GSM8K에서 gpt-3.5-turbo-0301 모델 이용해 3 agents and 2 rounds of debate를 벌임
  - 다만 선행 논문에서 100 example만 사용했던 것과 다르게 GSM8K 테스트셋 전체에서 테스트 수행
  - 참고용으로 self-consistency*도 수행  
\* prompts models to generate multiple responses and performs majority voting to select the final answer
- 실험 결과
  - Multi-agent debate나 Self-consistency 모두 Standard prompting보다는 나았다
  - 그런데 Agent 수를 동일하게 3개로 맞춰놓은 조건으로 비교하면 multi-agent debate가 self-consistency에 비해 아주 근소한 차이로 더 좋았고
  - 모댈 출력 개수를 동일하게 맞췄을 때는 multi-agent debate 성능이 오히려 낮았다
- 사실 Multi-agent debate는 실제로 토론을 한다기보단 그냥 consistency를 맞추기 위해 generation을 여러 번 돌린다고 표현하는 게 더 정확할 것임
- 결국 self-consistency의 기본 개념과 동일 선상인데, 둘 간의 차이라고 하면 majority voting의 메커니즘의 차이인데 Multi-agent debate는 model-driven이고 self-consistency는 그냥 카운트 기반인 것
- 그래서 지금까지의 관찰 결과로 봤을 때 응답의 개선은 self-correction보다는 self-consistency에서 비롯됐고, 따라서 LLM의 self-correct 능력에 대해서 논하려면 여러 응답 중에서 하나 고르기에서 오는 효과는 제하고 논해야 한다

## Self-Correction as Post-Hoc Prompting

- 그래도 Self-correction 자체는 다른 연구들에서는 놀라운 결과를 보였다.
- 일종의 post-hoc prompting으로써 prompting을 LLM의 출력을 베이스로 수행한다는 점에서 standard prompting (일종의 pre-hoc prompting)과 차이가 있다 
- 근데 모델 응답의 harmness를 줄이는 태스크 같은 경우에는 pre-hoc prompting만 가지고는 부족할 수 있으니 post-hoc prompting을 통해 개선하는 게 가능하지만 Reasoning task에는 부적절할 수도 있음
- 또 post-hoc prompting을 통해 개선된 사례도 사실 prompt에 명확한 요구사항이 있어서 pre-hoc prompting만으로도 충분한 효과를 낼 수 있었던 건 아닌지 분석해볼 필요가 있음
- 실제로 선행 논문 중 Constrained generation task에 대해 standard prompting하고 Self-correction을 비교한 논문이 있었는데 모델은 재현이 불가능해서 gpt-3.5-turbo-0613으로 설정해두고 그 논문에서 사용한 prompt를 사용한 결과랑 우리가 생각했을 때 가장 이상적인 initial prompt를 넣었을 때의 실험 결과를 비교해보니 standard prompting이 좀 압도적으로 좋았다
- 그래서 결과적으로 하고 싶은 말은 self-correction에 대해 분석을 할 때는 좀 더 비판적인 자세로 봐야지 않나, 최소한 pre-hoc과 post-hoc prompt 둘 다 동등한 노력을 들여서 만든 다음에 비교를 해야 공정한 비교가 될 것으로 보인다

## Discussion

- Self-correction may still be beneficial for aligning responses with certain preferences
  - Style transfer나 harmness 줄이기에는 도움이 된다
  - 하지만 LLM 스스로가 최초 응답에 대해 스스로 평가하기 어려운 태스크에 대해서는 도움이 안 된다
- Leveraging external feedback for correction
  - search engine, calculator, ...
  - Code generation task에선 executor와 unit test를 활용하거나
  - High-quality dataset으로 학습한 verifier나 critique model 활용하기 등등
- Employing self-consistency as a method of self-verification
  - Oracle setting에서 성능이 좋다는 것은 그래도 정답이 LLM의 search space 내에 있긴 하다는 소리니 이걸 잘 활용하면 될 것이다
  - External feedback을 사용하거나
  - Self-consistency를 높이거나 (e.g. Tree of Thought)
- Pre-hoc vs post-hoc prompting
  - 앞서 언급했듯이(?) post-hoc prompting은 비용이나 레이턴시 면에서 다 나쁘다
  - 그러니 웬만하면 Pre-hoc prompting을 잘 설계해서 쓰자
  - 물론 post-hoc의 최대 이점은 external feedback임
- Guidelines for comparison
  - pre-hoc이랑 post-hoc 비교할 때 첫째로 inference cost 언급하고, 베이스라인과 동등한 number of calls/responses로 비교해라
  - 두 번째, pre-hoc은 거지같이 설계하고 post-hoc은 정교하게 프롬프트 설계한 다음에 비교하지 마라
  - 세 번째, external feedback의 효과를 설명할 때는 그 feedback이 적절한지 생각해라. 특히 response의 정확성을 판별할 때 ground-truth label을 쓰는 건 그냥 oracle이다.

## Conclusion

Large Language Models Cannot Self-Correct **"Reasoning"** **"Yet"**

## Limitations and Broader Impact

- 특정 벤치마크에 있어서는 reasoning performance를 증가시키는 프롬프팅 전략이 있긴 할 텐데, 그게 진짜 순수 intrinsic일지 아님 사실은 human insight나 training data가 섞여 들어가는 것과 같은 external feedback이 있었는지 주의할 필요가 있다
- 언제, 어떻게 self-correction이 효과를 발휘하는지에 대해서는 도메인 전문가들마저 명확하게 설명하지 못 하고 있으며, 몇몇 연구에서는 레이블 사용이라는 중요한 내용을 논문에서 잘 안 보이는 곳에 명시해둔다던지, 아니면 self-correction이라고 했는데 면밀히 따지고 보니 external feedback이었다 와 같은 사례가 이를 반증하고 있음
- 그럴 거면 그냥 self라는 용어를 빼고 correction with external tools/feedback이라고 명시해주자
